pallets/xyk/src/lib.rs

```diff
@@ -2533,7 +3001,204 @@ impl<T: Config> XykFunctionsTrait<T::AccountId> for Pallet<T> {
 			}
 		}
 
-		Ok(())
+		Ok(sold_asset_amount)
+	}
+
+	fn multiswap_buy_asset(
+		sender: T::AccountId,
+		swap_token_list: Vec<TokenId>,
+		bought_asset_amount: Self::Balance,
+		max_amount_in: Self::Balance,
+		err_upon_bad_slippage: bool,
+		err_upon_non_slippage_fail: bool,
+	) -> Result<Balance, DispatchError> {
+		let (
+			fee_swap_buy_and_burn_amount,
+			fee_swap_treasury_amount,
+			fee_swap_pool_fee_amount,
+			fee_swap_input_reserve,
+			fee_swap_output_reserve,
+			fee_swap_sold_asset_id,
+			fee_swap_bought_asset_id,
+		) = <Pallet<T> as PreValidateSwaps>::pre_validate_multiswap_buy_asset(
+			&sender,
+			swap_token_list.clone(),
+			bought_asset_amount,
+			max_amount_in,
+		)?;
+
+		// First execute all atomic swaps in a storage layer
+		// And then the finally sold amount is compared
+		// The bool in error represents if the fail is due to bad final slippage
+		let res_multiswap =
```

Comment:
code style comment:

Let's move the whole functionality of let res_multiswap = calculation to a separate method.